llvm/lib/Transforms/InstCombine/InstCombineCasts.cpp

```diff
@@ -997,7 +997,8 @@ Instruction *InstCombinerImpl::transformZExtICmp(ICmpInst *Cmp,
         X = Builder.CreateNot(X);
       Value *Lshr = Builder.CreateLShr(X, ShAmt);
       Value *And1 = Builder.CreateAnd(Lshr, ConstantInt::get(X->getType(), 1));
-      return replaceInstUsesWith(Zext, And1);
+      return replaceInstUsesWith(
+          Zext, Builder.CreateZExtOrTrunc(And1, Zext.getType()));
```

Comment:
I think in the ICMP_EQ case, we may also have to check that the shl is one use now? Otherwise we replace zext+icmp+and with not+lshr+and+zext/trunc, which is more instructions.

Reply:
Right. I've added a negative test and adjusted the implementation accordingly.